File: too_many_cooks_data/test/registration_test.dart
/// Tests for agent registration.
library;

import 'dart:js_interop';

import 'package:dart_node_core/dart_node_core.dart';
import 'package:nadz/nadz.dart';
import 'package:test/test.dart';
import 'package:too_many_cooks_data/too_many_cooks_data.dart';

extension type _Fs(JSObject _) implements JSObject {
  external void unlinkSync(String path);
  external bool existsSync(String path);
}

final _Fs _fs = _Fs(requireModule('fs') as JSObject);

void _deleteIfExists(String path) {
  try {
    if (_fs.existsSync(path)) {
      _fs.unlinkSync(path);
    }
  } on Object catch (_) {}
}

void main() {
  const testDbPath = '.test_registration.db';
  TooManyCooksDb? db;

  setUp(() {
    _deleteIfExists(testDbPath);
    final config = createDataConfig(dbPath: testDbPath);
    final result = createDb(config);
    expect(result, isA<Success<TooManyCooksDb, String>>());
    db = (result as Success<TooManyCooksDb, String>).value;
  });

  tearDown(() {
    db?.close();
    _deleteIfExists(testDbPath);
  });

  test('register creates agent with key', () {
    final result = db!.register('test-agent');
    expect(result, isA<Success<AgentRegistration, DbError>>());
    final reg = (result as Success<AgentRegistration, DbError>).value;
    expect(reg.agentName, 'test-agent');
    expect(reg.agentKey.length, 64);
  });

  test('register fails for duplicate name', () {
    db!.register('duplicate-agent');
    final result = db!.register('duplicate-agent');
    expect(result, isA<Error<AgentRegistration, DbError>>());
    final error = (result as Error<AgentRegistration, DbError>).error;
    expect(error.code, errValidation);
    expect(error.message, contains('already registered'));
  });

  test('register fails for empty name', () {
    final result = db!.register('');
    expect(result, isA<Error<AgentRegistration, DbError>>());
    final error = (result as Error<AgentRegistration, DbError>).error;
    expect(error.code, errValidation);
    expect(error.message, contains('1-50'));
  });

  test('register fails for name over 50 chars', () {
    final result = db!.register('a' * 51);
    expect(result, isA<Error<AgentRegistration, DbError>>());
    final error = (result as Error<AgentRegistration, DbError>).error;
    expect(error.code, errValidation);
    expect(error.message, contains('1-50'));
  });

  test('register accepts name of exactly 50 chars', () {
    final result = db!.register('a' * 50);
    expect(result, isA<Success<AgentRegistration, DbError>>());
  });

  test('listAgents returns registered agents', () {
    db!.register('agent1');
    db!.register('agent2');
    final result = db!.listAgents();
    expect(result, isA<Success<List<AgentIdentity>, DbError>>());
    final agents = (result as Success<List<AgentIdentity>, DbError>).value;
    expect(agents.length, 2);
    expect(agents.map((a) => a.agentName).toSet(), {'agent1', 'agent2'});
  });
}

File: too-many-cooks/test/registration_test.ts
/// Tests for agent registration.

import { describe, it, beforeEach, afterEach } from "node:test";
import assert from "node:assert";
import fs from "node:fs";
import {
  type TooManyCooksDb,
  createDb,
  createDataConfig,
  ERR_VALIDATION,
} from "../lib/src/data/data.js";

const TEST_DB_PATH = ".test_registration.db";

const deleteIfExists = (path: string): void => {
  try {
    if (fs.existsSync(path)) {
      fs.unlinkSync(path);
    }
  } catch {
    // ignore
  }
};

describe("registration", () => {
  let db: TooManyCooksDb | undefined;

  beforeEach(() => {
    deleteIfExists(TEST_DB_PATH);
    const config = createDataConfig({ dbPath: TEST_DB_PATH });
    const result = createDb(config);
    assert.strictEqual(result.ok, true);
    if (!result.ok) {throw new Error("expected ok");}
    db = result.value;
  });

  afterEach(() => {
    db?.close();
    deleteIfExists(TEST_DB_PATH);
  });

  it("register creates agent with key", () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    const result = db.register("test-agent");
    assert.strictEqual(result.ok, true);
    if (!result.ok) {throw new Error("expected ok");}
    const reg = result.value;
    assert.strictEqual(reg.agentName, "test-agent");
    assert.strictEqual(reg.agentKey.length, 64);
  });

  it("register fails for duplicate name", () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    db.register("duplicate-agent");
    const result = db.register("duplicate-agent");
    assert.strictEqual(result.ok, false);
    if (result.ok) {throw new Error("expected error");}
    assert.strictEqual(result.error.code, ERR_VALIDATION);
    assert.ok(result.error.message.includes("already registered"));
  });

  it("register fails for empty name", () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    const result = db.register("");
    assert.strictEqual(result.ok, false);
    if (result.ok) {throw new Error("expected error");}
    assert.strictEqual(result.error.code, ERR_VALIDATION);
    assert.ok(result.error.message.includes("1-50"));
  });

  it("register fails for name over 50 chars", () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    const result = db.register("a".repeat(51));
    assert.strictEqual(result.ok, false);
    if (result.ok) {throw new Error("expected error");}
    assert.strictEqual(result.error.code, ERR_VALIDATION);
    assert.ok(result.error.message.includes("1-50"));
  });

  it("register accepts name of exactly 50 chars", () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    const result = db.register("a".repeat(50));
    assert.strictEqual(result.ok, true);
  });

  it("listAgents returns registered agents", () => {
    assert.notStrictEqual(db, undefined);
    if (!db) {throw new Error("expected db");}
    db.register("agent1");
    db.register("agent2");
    const result = db.listAgents();
    assert.strictEqual(result.ok, true);
    if (!result.ok) {throw new Error("expected ok");}
    const agents = result.value;
    assert.strictEqual(agents.length, 2);
    assert.deepStrictEqual(
      new Set(agents.map((a) => a.agentName)),
      new Set(["agent1", "agent2"]),
    );
  });
});
